File: Cpp/Sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>


void Merge(std::vector<int>& A, std::vector<int>& result, int left, int mid, int right) {
    int i = 0;
    int j = 0;

    while (left + i < mid && mid + j < right) {
        if (A[left + i] < A[mid + j]) {
            result[i + j] = A[left + i];
            i++;
        }
        else {
            result[i + j] = A[mid + j];
            j++;
        }
    }

    while (left + i < mid) {
        result[i + j] = A[left + i];
        i++;
    }

    while (mid + j < right) {
        result[i + j] = A[mid + j];
        j++;
    }

    for (int step = 0; step < i + j; step++)
        A[left + step] = result[step];
}

void Sort(std::vector<int>& A) {
    int size = A.size();
    std::vector<int> result(size);
    for (int i = 1; i < size; i *= 2) {
        for (int j = 0; j < size - i; j += 2 * i) {
            Merge(A, result,  j, j + i, std::min(j + 2 * i, size));
        }
    }
}

int main() {
    setlocale(0, "");
    int size = 0;

    std::cout << "Specify the size of the array: " << std::endl;
    std::cin >> size;
    std::vector<int> A(size);

    std::cout << "Enter array elements separated by a space: " << std::endl;
    for (int i = 0; i < size; i++)
        std::cin >> A[i];

    std::cout << "Array size: " << size << std::endl;
    for (int i = 0; i < size - 1; i++)
        std::cout << A[i] << " ";
    std::cout << A[size - 1] << std::endl;

    Sort(A);

    std::cout << "Sort result:" << std::endl;
    for (int i = 0; i < size; i++)
        std::cout << A[i] << " ";

    return 0;
}

File: Python/Sort.py


def Merge(array, res_array, left, mid, right):
    i = 0
    j = 0
    while (left + i < mid and mid + j < right):
        if (array[left + i] < array[mid + j]):
            res_array[i + j] = array[left + i]
            i += 1
        else:
            res_array[i + j] = array[mid + j]
            j += 1

    while (left + i < mid):
        res_array[i + j] = array[left + i]
        i += 1

    while (mid + j < right):
        res_array[i + j] = array[mid + j]
        j += 1

    for step in range(0, i + j):
        array[left + step] = res_array[step];

def Sort(array):
    size = len(array)
    result = [0 for i in range(size)]
    i = 1
    while i < size:
        j = 0
        while j < size - i:
            Merge(array, result,  j, j + i, min(j + 2 * i, size))
            j += 2 * i
        i *= 2


size = int(input("Укажите размер массива: "));
A = [0 for i in range(size)]

array = list(map(int, input("Введите элементы массива через пробел: ").split()))

print("Исходный массив размера ", size)
print(array)

Sort(array)

print("Результат сортировки:")
print(array)

